tokio-util/tests/future.rs

```diff
@@ -0,0 +1,255 @@
+    let res = fut.as_mut().poll(&mut Context::from_waker(&waker));
+    assert_ready_eq!(res, Some(()));
+
+    let res = fut.as_mut().poll(&mut Context::from_waker(&waker));
+    assert_ready_eq!(res, Some(()));
+
+    let res = fut.as_mut().poll(&mut Context::from_waker(&waker));
+    assert_ready_eq!(res, Some(()));
+}
+
+#[test]
+fn ready_future_with_cancellation_owned_token_test() {
+    let (waker, _) = new_count_waker();
+    let token = CancellationToken::new();
+
+    let ready_fut = ready(());
+
+    let ready_with_token_fut = ready_fut.with_cancellation_token_owned(token);
+
+    pin!(ready_with_token_fut);
+
+    let res = ready_with_token_fut
+        .as_mut()
+        .poll(&mut Context::from_waker(&waker));
+
+    assert_ready_eq!(res, Some(()));
+}
+
+#[test]
+fn pending_future_with_cancellation_token_owned_test() {
+    let (waker, _) = new_count_waker();
+    let token = CancellationToken::new();
+
+    let pending_fut = pending::<()>();
+
+    let pending_with_token_fut = pending_fut.with_cancellation_token_owned(token);
+
+    pin!(pending_with_token_fut);
+
+    let res = pending_with_token_fut
+        .as_mut()
+        .poll(&mut Context::from_waker(&waker));
+
+    assert_pending!(res);
+}
+
+#[test]
+fn pending_fut_with_already_cancelled_token_owned_test() {
+    let (waker, wake_count) = new_count_waker();
+    let token = CancellationToken::new();
+    token.cancel();
+
+    let pending_fut = pending::<()>();
+
+    let pending_with_token_fut = pending_fut.with_cancellation_token_owned(token);
+
+    pin!(pending_with_token_fut);
+
+    let res = pending_with_token_fut
+        .as_mut()
+        .poll(&mut Context::from_waker(&waker));
+
+    assert_ready_eq!(res, None);
+    assert_eq!(wake_count, 0);
+}
+
+#[test]
+fn pending_fut_with_owned_token_cancelled_test() {
+    let (waker, wake_count) = new_count_waker();
+    let token = CancellationToken::new();
+
+    let pending_fut = pending::<()>();
+
+    let pending_with_token_fut = pending_fut.with_cancellation_token_owned(token.clone());
+
+    pin!(pending_with_token_fut);
+
+    let res = pending_with_token_fut
+        .as_mut()
+        .poll(&mut Context::from_waker(&waker));
+    assert_pending!(res);
+
+    token.cancel();
+
+    let res = pending_with_token_fut
+        .as_mut()
+        .poll(&mut Context::from_waker(&waker));
+    assert_ready_eq!(res, None);
+    assert_eq!(wake_count, 1);
+}
+
+#[test]
+fn pending_only_on_first_poll_with_cancellation_token_owned_test() {
+    #[derive(Default)]
+    struct TestFuture {
+        polled: bool,
+    }
+
+    impl Future for TestFuture {
+        type Output = ();
+
+        fn poll(mut self: std::pin::Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+            if !self.polled {
+                self.polled = true;
+                return Poll::Pending;
+            }
+            Poll::Ready(())
+        }
+    }
```

Comment:
nit: We'd better avoiding the duplicated implementation. We also want a better name like `ReadyOnTheSecondPollFuture`.